tmpl/blog/introducing-charrua-dhcp.md

```diff
@@ -0,0 +1,221 @@
+
+val input_pkt : Dhcp_server.Config.t ->  Dhcp_server.Config.subnet ->
+   Dhcp_wire.pkt -> float -> result
+(** input_pkt config subnet pkt time Inputs packet pkt, the resulting action
+    should be performed by the caller, normally a Reply packet is returned and
+    must be sent on the same subnet. time is a float representing time as in
+    Unix.time or MirageOS's Clock.time. **)
+```
+
+A typical main server loop would work by:
+ 1. Reading a packet from the network.
+ 2. Unmarshaling with [Dhcp_wire.pkt_of_buf]
+(http://haesbaert.github.io/charrua-core/api/Dhcp_wire.html#VALpkt_of_buf).
+ 3. Inputing the result with [Dhcp_server.Input.input_pkt]
+(http://haesbaert.github.io/charrua-core/api/Dhcp_server.Input.html#VALinput_pkt).
+ 4. Sending the reply, or logging the event from the [Dhcp_server.Input.input_pkt]
+(http://haesbaert.github.io/charrua-core/api/Dhcp_server.Input.html#VALinput_pkt) call.
+
+A mainloop example can be found in
+[mirage-skeleton](https://github.com/mirage/mirage-skeleton/blob/master/dhcp/unikernel.ml#L28):
+
+```ocaml
+  let input_dhcp c net config subnet buf =
+    let open Dhcp_server.Input in
+    match (Dhcp_wire.pkt_of_buf buf (Cstruct.len buf)) with
+    | `Error e -> Lwt.return (log c (red "Can't parse packet: %s" e))
+    | `Ok pkt ->
+      match (input_pkt config subnet pkt (Clock.time ())) with
+      | Silence -> Lwt.return_unit
+      | Warning w -> Lwt.return (log c (yellow "%s" w))
+      | Error e -> Lwt.return (log c (red "%s" e))
+      | Reply reply ->
+        log c (blue "Received packet %s" (Dhcp_wire.pkt_to_string pkt));
+        N.write net (Dhcp_wire.buf_of_pkt reply)
+        >>= fun () ->
+        log c (blue "Sent reply packet %s" (Dhcp_wire.pkt_to_string reply));
+        Lwt.return_unit
+```
+
+As stated,
+[Dhcp_server.Input.input_pkt] (http://haesbaert.github.io/charrua-core/api/Dhcp_server.Input.html#VALinput_pkt)
+does not perform any IO of its own, it deals only with the logic of inputing a
+DHCP packet and returning a possible answer, which should then be sent by the
+caller. This allows a design where all the side effects are controlled in one
+small chunk, it makes it easier to understand the state transitions when they
+are made explicit.
+
+At the time of this writing, [Dhcp_server.Input.input_pkt]
+(http://haesbaert.github.io/charrua-core/api/Dhcp_server.Input.html#VALinput_pkt)
+is not side effect free, as it manipulates a database of leases, this will be
+changed in the next version to be pure as well.
+
+Storing leases in permanent storage is also not supported at this time and
+should be available soon, with irmin and other backends. The main idea is to
+always return a new lease database at every input, or maybe just the updates to
+be applied, in this scenario, the caller would be able to store the database in
+permanent storage as he sees fit.
+
+#### Configuration
+
+The project started before the integration with MirageOS, and at that time, the
```

Comment:
Do you mean: "This project started independently of MirageOS" ? It's not clear to me what 'before the integration with MirageOS' means otherwise.